### PR TITLE
Make span.Name settable, update it with the route info for ASP.NET Core

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Make name settable https://github.com/census-instrumentation/opencensus-specs/issues/217 and use it

### CODE REVIEW

File: src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs [hunk 1]
```diff
@@ -31,6 +30,9 @@ internal class HttpInListener : ListenerHandler
         private const string UnknownHostName = "UNKNOWN-HOST";
         private readonly PropertyFetcher startContextFetcher = new PropertyFetcher("HttpContext");
         private readonly PropertyFetcher stopContextFetcher = new PropertyFetcher("HttpContext");
+        private readonly PropertyFetcher beforActionActionDescriptorFetcher = new PropertyFetcher("actionDescriptor");
```
Comment (simonz130) [inline src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs]: nit: before

File: src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs [hunk 2]
```diff
@@ -101,6 +103,30 @@ public override void OnStopActivity(Activity activity, object payload)
...
 
+        public override void OnCustom(string name, Activity activity, object payload)
+        {
+            if (name == "Microsoft.AspNetCore.Mvc.BeforeAction")
+            {
+                var span = this.Tracer.CurrentSpan;
+
+                if (span == null)
+                {
+                    // report lost span
```
Comment (simonz130) [inline src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs]: add todo

File: src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs [hunk 3]
```diff
@@ -101,6 +103,30 @@ public override void OnStopActivity(Activity activity, object payload)
             span.End();
         }
 
+        public override void OnCustom(string name, Activity activity, object payload)
+        {
+            if (name == "Microsoft.AspNetCore.Mvc.BeforeAction")
```
Comment (simonz130) [inline src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs]: can you do typeof(Microsoft.AspNetCore.Mvc.BeforeAction).Name?

File: src/OpenCensus.Collector.AspNetCore/RequestsCollector.cs [hunk 4]
```diff
@@ -41,7 +41,9 @@ public RequestsCollector(RequestsCollectorOptions options, ITracer tracer, ISamp
         {
             this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(
                 new Dictionary<string, Func<ITracer, ISampler, ListenerHandler>>()
-                { { "Microsoft.AspNetCore", (t, s) => new HttpInListener(t, s, propagationComponent) } },
+                {
+                    { "Microsoft.AspNetCore", (t, s) => new HttpInListener(t, s, propagationComponent) },
```
Comment (simonz130) [inline src/OpenCensus.Collector.AspNetCore/RequestsCollector.cs]: can you reference a type as opposed to hardcoded string?

File: src/OpenCensus/Trace/Span.cs [hunk 5]
```diff
@@ -77,12 +77,18 @@ public sealed class Span : SpanBase
...
             get
             {
                 return this.name;
             }
+
+            set
+            {
+                this.name = value;
+            }
         }
```
Comment (simonz130) [inline src/OpenCensus/Trace/Span.cs]: Why not ```public override string Name { get;set; }```?

File: src/OpenCensus/Trace/Span.cs [hunk 6]
```diff
@@ -77,12 +77,18 @@ public sealed class Span : SpanBase
...
         public override string Name
         {
             get
             {
                 return this.name;
             }
+
+            set
+            {
+                this.name = value;
```
Comment (simonz130) [inline src/OpenCensus/Trace/Span.cs]: maybe it's a style preference, but why adding "this." where it's not required?

File: src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs (hunk 2 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs]: sone

File: src/OpenCensus.Collector.AspNetCore/RequestsCollector.cs (hunk 4 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.AspNetCore/RequestsCollector.cs]: No. One of ideas of DiagnosticsSource is ducktyping and eliminating direct referencens between packages. I added comment where I use reflection

File: src/OpenCensus/Trace/Span.cs (hunk 5 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus/Trace/Span.cs]: yes, done. It will make code simpler

File: src/OpenCensus/Trace/Span.cs (hunk 6 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus/Trace/Span.cs]: it's stylecop rule. There are two major camps: `_name` for fields and `this.name` for fields. I just use whatever default in Stylecop, which is `this.name`

**Comment by SergeyKanzhelev** [comment]
@simonz130 addressed all feedback. Thank you for review!

File: src/OpenCensus.Abstractions/Trace/ISampler.cs [hunk 7]
```diff
@@ -35,7 +35,11 @@ public interface ISampler
         /// <param name="hasRemoteParent">Indicates whether it was a remote parent.</param>
         /// <param name="traceId">Trace ID of a span to be created.</param>
         /// <param name="spanId">Span ID of a span to be created.</param>
-        /// <param name="name">Name of a span to be created.</param>
+        /// <param name="name"> Name of a span to be created. Note, that the name of the span is settable.
+        /// So this name can be changed later and <see cref="ISampler"/> implementation should assume that.
+        /// Typical example of a name change is when <see cref="ISpan"/> representing incoming http request
+        /// has a name of url path and then being updated with route name when rouing complete.
```
Comment (lmolkova) [inline src/OpenCensus.Abstractions/Trace/ISampler.cs]: why do we do this through span name and not through the `http.route` attribute?

File: src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs [hunk 8]
```diff
@@ -101,6 +103,33 @@ public override void OnStopActivity(Activity activity, object payload)
...
+            {
+                var span = this.Tracer.CurrentSpan;
+
+                if (span == null)
+                {
+                    // TODO: report lost span
```
Comment (lmolkova) [inline src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs]: and return or span.Name will throw

File: src/OpenCensus.Collector.Implementation.Common/PropertyFetcher.cs [hunk 9]
```diff
@@ -33,7 +33,14 @@ public object Fetch(object obj)
         {
             if (this.innerFetcher == null)
             {
-                this.innerFetcher = PropertyFetch.FetcherForProperty(obj.GetType().GetTypeInfo().GetDeclaredProperty(this.propertyName));
+                var type = obj.GetType().GetTypeInfo();
+                var property = type.GetDeclaredProperty(this.propertyName);
+                if (property == null)
+                {
+                    property = type.GetProperty(this.propertyName);
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Implementation.Common/PropertyFetcher.cs]: are we doing private reflection here?

File: src/OpenCensus.Abstractions/Trace/ISampler.cs (hunk 7 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Abstractions/Trace/ISampler.cs]: span name should have low cardinality. Thus we set route into name. Perhaps we also need to put route into the attribute.

File: src/OpenCensus.Collector.Implementation.Common/PropertyFetcher.cs (hunk 9 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Implementation.Common/PropertyFetcher.cs]: It's not private reflection. It's getting property of base class

**Pull request merged**